File: sabot/query/logical_plan.py
#!/usr/bin/env python3
"""
Unified Logical Plan

Logical query representation that all APIs (Stream, SQL, Graph) translate to.
Enables shared optimization and composition.

Design: Similar to DuckDB/DataFusion logical plans but simpler.
"""

import logging
from typing import Optional, List, Any, Dict
from dataclasses import dataclass
from enum import Enum
from abc import ABC, abstractmethod

logger = logging.getLogger(__name__)


class LogicalOperatorType(Enum):
    """Types of logical operators."""
    # Data sources
    SCAN = "scan"
    
    # Transforms
    FILTER = "filter"
    PROJECT = "project"
    MAP = "map"
    
    # Aggregations
    AGGREGATE = "aggregate"
    GROUP_BY = "group_by"
    
    # Joins
    JOIN = "join"
    
    # Windows
    WINDOW = "window"
    
    # Sorting/Limiting
    SORT = "sort"
    LIMIT = "limit"
    
    # Set operations
    UNION = "union"
    DISTINCT = "distinct"


class LogicalPlan(ABC):
    """
    Base class for logical plan nodes.
    
    All query APIs (Stream, SQL, Graph) translate to logical plans,
    enabling shared optimization and composition.
    
    Design principles:
    - Immutable (functional transformations)
    - Tree structure (parent-child relationships)
    - Type-safe (strong typing)
    - Serializable (can be sent over network)
    """
    
    def __init__(self, operator_type: LogicalOperatorType):
        self.operator_type = operator_type
        self.children: List[LogicalPlan] = []
        self.metadata: Dict[str, Any] = {}
    
    @abstractmethod
    def to_dict(self) -> Dict[str, Any]:
        """Serialize to dictionary."""
        pass
    
    @abstractmethod
    def validate(self) -> bool:
        """Validate plan correctness."""
        pass
    
    def add_child(self, child: 'LogicalPlan'):
        """Add child operator."""
        self.children.append(child)
    
    def get_children(self) -> List['LogicalPlan']:
        """Get child operators."""
        return self.children


@dataclass
class ScanPlan(LogicalPlan):
    """
    Scan data source.
    
    Represents reading from:
    - Files (Parquet, CSV, Arrow IPC)
    - Kafka topics
    - Database tables
    - In-memory tables
    """
    source_type: str  # 'kafka', 'parquet', 'table', etc.
    source_name: str  # Topic name, file path, table name
    schema: Optional[Any] = None
    options: Optional[Dict[str, Any]] = None
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.SCAN)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'scan',
            'source_type': self.source_type,
            'source_name': self.source_name,
            'options': self.options or {}
        }
    
    def validate(self) -> bool:
        return bool(self.source_name)


@dataclass
class FilterPlan(LogicalPlan):
    """
    Filter rows based on predicate.
    
    Example:
        FilterPlan(input=scan, predicate="x > 10")
    """
    input: LogicalPlan
    predicate: Any  # Can be string expression or callable
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.FILTER)
        self.add_child(self.input)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'filter',
            'predicate': str(self.predicate),
            'input': self.input.to_dict()
        }
    
    def validate(self) -> bool:
        return self.input is not None and self.input.validate()


@dataclass
class ProjectPlan(LogicalPlan):
    """
    Project (select) columns.
    
    Example:
        ProjectPlan(input=scan, columns=['a', 'b', 'c'])
    """
    input: LogicalPlan
    columns: List[str]
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.PROJECT)
        self.add_child(self.input)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'project',
            'columns': self.columns,
            'input': self.input.to_dict()
        }
    
    def validate(self) -> bool:
        return bool(self.columns) and self.input.validate()


@dataclass
class JoinPlan(LogicalPlan):
    """
    Join two inputs.
    
    Example:
        JoinPlan(left=scan1, right=scan2, left_keys=['id'], right_keys=['user_id'], join_type='inner')
    """
    left: LogicalPlan
    right: LogicalPlan
    left_keys: List[str]
    right_keys: List[str]
    join_type: str = 'inner'  # 'inner', 'left', 'right', 'outer', 'semi', 'anti'
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.JOIN)
        self.add_child(self.left)
        self.add_child(self.right)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'join',
            'join_type': self.join_type,
            'left_keys': self.left_keys,
            'right_keys': self.right_keys,
            'left': self.left.to_dict(),
            'right': self.right.to_dict()
        }
    
    def validate(self) -> bool:
        return (
            self.left.validate() and
            self.right.validate() and
            len(self.left_keys) == len(self.right_keys)
        )


@dataclass
class GroupByPlan(LogicalPlan):
    """
    Group by keys and aggregate.
    
    Example:
        GroupByPlan(input=scan, group_keys=['category'], aggregations={'amount': 'sum'})
    """
    input: LogicalPlan
    group_keys: List[str]
    aggregations: Dict[str, str]  # column -> function ('sum', 'count', 'avg', etc.)
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.GROUP_BY)
        self.add_child(self.input)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'group_by',
            'group_keys': self.group_keys,
            'aggregations': self.aggregations,
            'input': self.input.to_dict()
        }
    
    def validate(self) -> bool:
        return bool(self.group_keys) and bool(self.aggregations) and self.input.validate()


@dataclass
class WindowPlan(LogicalPlan):
    """
    Windowed aggregation.
    
    Example:
        WindowPlan(input=scan, window_config=tumbling(60.0))
    """
    input: LogicalPlan
    window_config: Any  # WindowConfig from window_unified
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.WINDOW)
        self.add_child(self.input)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'window',
            'window_type': self.window_config.window_type.value,
            'size': self.window_config.size,
            'input': self.input.to_dict()
        }
    
    def validate(self) -> bool:
        return self.window_config is not None and self.input.validate()


@dataclass
class LimitPlan(LogicalPlan):
    """
    Limit number of rows.
    
    Example:
        LimitPlan(input=scan, limit=100, offset=0)
    """
    input: LogicalPlan
    limit: int
    offset: int = 0
    
    def __post_init__(self):
        super().__init__(LogicalOperatorType.LIMIT)
        self.add_child(self.input)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'type': 'limit',
            'limit': self.limit,
            'offset': self.offset,
            'input': self.input.to_dict()
        }
    
    def validate(self) -> bool:
        return self.limit > 0 and self.input.validate()


class LogicalPlanBuilder:
    """
    Builder for constructing logical plans.
    
    Provides fluent API for building plans programmatically.
    
    Example:
        plan = (LogicalPlanBuilder()
            .scan('kafka', 'transactions')
            .filter('amount > 1000')
            .project(['id', 'amount', 'customer_id'])
            .group_by(['customer_id'], {'amount': 'sum'})
            .build())
    """
    
    def __init__(self):
        self._plan: Optional[LogicalPlan] = None
    
    def scan(self, source_type: str, source_name: str, **options) -> 'LogicalPlanBuilder':
        """Add scan operator."""
        self._plan = ScanPlan(source_type, source_name, options=options)
        return self
    
    def filter(self, predicate: Any) -> 'LogicalPlanBuilder':
        """Add filter operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = FilterPlan(self._plan, predicate)
        return self
    
    def project(self, columns: List[str]) -> 'LogicalPlanBuilder':
        """Add project operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = ProjectPlan(self._plan, columns)
        return self
    
    def group_by(self, keys: List[str], aggregations: Dict[str, str]) -> 'LogicalPlanBuilder':
        """Add group by operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = GroupByPlan(self._plan, keys, aggregations)
        return self
    
    def join(
        self,
        right: LogicalPlan,
        left_keys: List[str],
        right_keys: List[str],
        join_type: str = 'inner'
    ) -> 'LogicalPlanBuilder':
        """Add join operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = JoinPlan(self._plan, right, left_keys, right_keys, join_type)
        return self
    
    def window(self, window_config: Any) -> 'LogicalPlanBuilder':
        """Add window operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = WindowPlan(self._plan, window_config)
        return self
    
    def limit(self, limit: int, offset: int = 0) -> 'LogicalPlanBuilder':
        """Add limit operator."""
        if self._plan is None:
            raise ValueError("Must call scan() first")
        self._plan = LimitPlan(self._plan, limit, offset)
        return self
    
    def build(self) -> LogicalPlan:
        """Build and return logical plan."""
        if self._plan is None:
            raise ValueError("Empty plan - must add operators")
        
        if not self._plan.validate():
            raise ValueError("Invalid plan")
        
        return self._plan

File: sabot_core/src/query/logical_plan.cpp
#include "sabot/query/logical_plan.h"

namespace sabot {
namespace query {

// ScanPlan implementation

ScanPlan::ScanPlan(const std::string& source_type, const std::string& source_name)
    : source_type_(source_type), source_name_(source_name), estimated_rows_(-1) {}

bool ScanPlan::Validate() const {
    return !source_name_.empty();
}

arrow::Result<std::shared_ptr<arrow::Schema>> ScanPlan::GetSchema() const {
    if (schema_) {
        return schema_;
    }
    
    // Schema unknown for now
    return arrow::Status::NotImplemented("Schema inference not implemented");
}

int64_t ScanPlan::EstimateCardinality() const {
    return estimated_rows_ > 0 ? estimated_rows_ : 1000000;  // Default estimate
}

std::unordered_map<std::string, std::string> ScanPlan::GetMetadata() const {
    return {
        {"operator", "scan"},
        {"source_type", source_type_},
        {"source_name", source_name_}
    };
}

std::shared_ptr<LogicalPlan> ScanPlan::Clone() const {
    auto clone = std::make_shared<ScanPlan>(source_type_, source_name_);
    clone->schema_ = schema_;
    clone->estimated_rows_ = estimated_rows_;
    return clone;
}

// FilterPlan implementation

FilterPlan::FilterPlan(std::shared_ptr<LogicalPlan> input, const std::string& predicate)
    : input_(input), predicate_(predicate), selectivity_(0.5) {}

bool FilterPlan::Validate() const {
    return input_ && input_->Validate() && !predicate_.empty();
}

arrow::Result<std::shared_ptr<arrow::Schema>> FilterPlan::GetSchema() const {
    // Filter doesn't change schema
    return input_->GetSchema();
}

int64_t FilterPlan::EstimateCardinality() const {
    // Estimate based on selectivity
    int64_t input_rows = input_->EstimateCardinality();
    return static_cast<int64_t>(input_rows * selectivity_);
}

std::unordered_map<std::string, std::string> FilterPlan::GetMetadata() const {
    return {
        {"operator", "filter"},
        {"predicate", predicate_},
        {"selectivity", std::to_string(selectivity_)}
    };
}

std::shared_ptr<LogicalPlan> FilterPlan::Clone() const {
    auto clone = std::make_shared<FilterPlan>(input_->Clone(), predicate_);
    clone->selectivity_ = selectivity_;
    return clone;
}

// ProjectPlan implementation

ProjectPlan::ProjectPlan(
    std::shared_ptr<LogicalPlan> input,
    const std::vector<std::string>& columns
) : input_(input), columns_(columns) {}

bool ProjectPlan::Validate() const {
    return input_ && input_->Validate() && !columns_.empty();
}

arrow::Result<std::shared_ptr<arrow::Schema>> ProjectPlan::GetSchema() const {
    // TODO: Build schema from selected columns
    return arrow::Status::NotImplemented("Project schema not implemented");
}

int64_t ProjectPlan::EstimateCardinality() const {
    // Projection doesn't change cardinality
    return input_->EstimateCardinality();
}

std::unordered_map<std::string, std::string> ProjectPlan::GetMetadata() const {
    std::string cols;
    for (size_t i = 0; i < columns_.size(); ++i) {
        if (i > 0) cols += ",";
        cols += columns_[i];
    }
    
    return {
        {"operator", "project"},
        {"columns", cols}
    };
}

std::shared_ptr<LogicalPlan> ProjectPlan::Clone() const {
    return std::make_shared<ProjectPlan>(input_->Clone(), columns_);
}

// JoinPlan implementation

JoinPlan::JoinPlan(
    std::shared_ptr<LogicalPlan> left,
    std::shared_ptr<LogicalPlan> right,
    const std::vector<std::string>& left_keys,
    const std::vector<std::string>& right_keys,
    JoinType join_type
) : left_(left), right_(right), left_keys_(left_keys), 
    right_keys_(right_keys), join_type_(join_type) {}

bool JoinPlan::Validate() const {
    return left_ && right_ && 
           left_->Validate() && right_->Validate() &&
           left_keys_.size() == right_keys_.size() &&
           !left_keys_.empty();
}

arrow::Result<std::shared_ptr<arrow::Schema>> JoinPlan::GetSchema() const {
    // TODO: Combine left and right schemas
    return arrow::Status::NotImplemented("Join schema not implemented");
}

int64_t JoinPlan::EstimateCardinality() const {
    // Simple estimate: multiply cardinalities and apply selectivity
    int64_t left_rows = left_->EstimateCardinality();
    int64_t right_rows = right_->EstimateCardinality();
    
    // Assume 10% join selectivity
    return static_cast<int64_t>(left_rows * right_rows * 0.1);
}

std::unordered_map<std::string, std::string> JoinPlan::GetMetadata() const {
    return {
        {"operator", "join"},
        {"join_type", std::to_string(static_cast<int>(join_type_))},
        {"left_keys", std::to_string(left_keys_.size())},
        {"right_keys", std::to_string(right_keys_.size())}
    };
}

std::shared_ptr<LogicalPlan> JoinPlan::Clone() const {
    return std::make_shared<JoinPlan>(
        left_->Clone(),
        right_->Clone(),
        left_keys_,
        right_keys_,
        join_type_
    );
}

// GroupByPlan implementation

GroupByPlan::GroupByPlan(
    std::shared_ptr<LogicalPlan> input,
    const std::vector<std::string>& group_keys,
    const std::unordered_map<std::string, std::string>& aggregations
) : input_(input), group_keys_(group_keys), aggregations_(aggregations) {}

bool GroupByPlan::Validate() const {
    return input_ && input_->Validate() && 
           !group_keys_.empty() && !aggregations_.empty();
}

arrow::Result<std::shared_ptr<arrow::Schema>> GroupByPlan::GetSchema() const {
    // TODO: Build schema from group keys + aggregations
    return arrow::Status::NotImplemented("GroupBy schema not implemented");
}

int64_t GroupByPlan::EstimateCardinality() const {
    // Estimate based on number of unique groups
    // Assume 10% unique groups
    int64_t input_rows = input_->EstimateCardinality();
    return static_cast<int64_t>(input_rows * 0.1);
}

std::unordered_map<std::string, std::string> GroupByPlan::GetMetadata() const {
    return {
        {"operator", "group_by"},
        {"num_keys", std::to_string(group_keys_.size())},
        {"num_aggs", std::to_string(aggregations_.size())}
    };
}

std::shared_ptr<LogicalPlan> GroupByPlan::Clone() const {
    return std::make_shared<GroupByPlan>(input_->Clone(), group_keys_, aggregations_);
}

} // namespace query
} // namespace sabot

